mantidproject/mantid
# Maintain aspect ratios in 2D instrument views

**Description of work.**

Add switch to the Display Settings dropdown on the Render tab to enable banks to be rendered using original aspect ratio without stretching to reflect size of plot area. This affects the following views in the instrument view:

- Cylindrical X, Y, Z
- Spherical X, Y, Z
- Side by Side

The Full 3D view already maintains the aspect ratio of the detectors regardless of the plot area dimensions. I guess I could have done the opposite change to this view ie enable the stretch to widget behaviour. But thought that might look a bit odd in a 3D view so have left it alone. So the new switch is only on the 2D views.

The motivation for the change is a request from SXD for some improvements to the side by side view where currently their square banks are stretched into rectangles by the somewhat arbitrary size of the plot widget in the instrument viewer. I've made the switch available in all the 2D views though because seems useful.

As part of this I have modified the definition of the Cylindrical X, Y, Z views so that they use a standard cylindrical projection (see UnwrappedCylinder.cpp). The code previously set u equal to the longitude of the detector (in radians) and v equal to the z coordinate of the detector. This mixed scheme of u being an angle and v being a distance meant the view was much wider than it was tall when the true aspect ratio was maintained. The new code uses an equal area cylindrical projection where v = sin(latitude) = sin(z/r)

**To test:**

Load run SXD33304.nxs. Right click on the resulting workspace and select "Show Instrument" and then select Side by Side View. It should look like this with the detectors displayed as squares (two of them are rotated) - the new "Maintain aspect ratio" option is on by default:

![image](https://user-images.githubusercontent.com/56295817/214883061-4c3f2731-96cd-4cf9-8d5f-b02cc824267c.png)

The detectors should remain as squares if you resize the window.
Go to the Pick tab and check the cursor readout responds as you move the cursor onto a bank from the black (ie blank) areas
Add a single crystal peak by clicking on the tool bar icon (see below) and then clicking on the instrument on an area of high intensity. Then click on a peak in the TOF plot on the LHS to finish adding the peak
![image](https://user-images.githubusercontent.com/56295817/215100992-1fc6a831-13a3-4934-a5fd-3128eb569168.png)
Check resizing the window keeps the peak displayed over the same detector(s)
Check the zoom feature works.
Go back to the Render tab and turn off the "Maintain aspect ratio" option and the banks should now be rectangular and they should stretch when you change the size of the window (original behaviour)

Check all the views on the dropdown (Full 3D, Cylindrical X etc) all display and that the new "Maintain aspect ratio" switch behaves as you'd expect as you change the size of the window. With the setting "on", not all of the views have the detectors filling the plot area but they shouldn't be ridiculously stretched in one dimension.

Finally open up the reflectometry UI and go to the Reduction Preview tab. Enter INTER45455_inst and click Load. The window has an embedded instrument view on the LHS which uses a cylindrical Y projection. Verify that this detector view stretches to fill the widget (it should be shown with "maintain aspect ratio" set to false to retain the original behaviour).
[INTER45455_inst.zip](https://github.com/mantidproject/mantid/files/10519226/INTER45455_inst.zip)

Fixes #34560 (partially)




---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
